### PR TITLE
container-collection: Handle OciConfig for initial containers

### PR DESCRIPTION
**Pull request by mqasimsarfraz**
This change add support for reading OciConfig for initial containers as well. 

Fixes: #3884 

### docker

```
$ make ig
$ make minikube-start-docker
$ docker cp ig minikube-docker:/bin/ig
$ minikube ssh 
$ sudo ig list-containers -o json
...
"ociConfig": "{\"ociVersion\":\"1.2.0\",\"process\":{\"user\":{\"uid\":0,\"gid\":0,\"additionalGids\":[0]},\"args\":[\"/storage-provisioner\"],\"env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\",\"HOSTNAME=minikube-docker\",\"KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443\",\"KUBE_DNS_SERVICE_HOST=10.96.0.10\",\"KUBE_DNS_SERVICE_PORT_DNS=53\",\"KUBE_DNS_PORT_53_UDP=udp://10.96.0.10:53\",\"KUBE_DNS_PORT_9153_TCP_PROTO=tcp\",\"KUBERNETES_SERVICE_HOST=10.96.0.1\",\"KUBERN
...
```

### containerd

```
$ make ig
$ make minikube-start-containerd
$ docker cp ig minikube-containerd:/bin/ig
$ minikube ssh 
$ sudo ig list-containers -o json
...
   "ociConfig": "{\"ociVersion\":\"1.2.0\",\"process\":{\"user\":{\"uid\":65532,\"gid\":65532,\"additionalGids\":[65532]},\"args\":[\"/coredns\",\"-conf\",\"/etc/coredns/Corefile\"],\"env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\",\"HOSTNAME=coredns-7db6d8ff4d-m5sfp\",\"KUBE_DNS_SERVICE_PORT_DNS_TCP=53\",\"KUBE_DNS_SERVICE_PORT_METRICS=9153\",\"KUBE_DNS_PORT_53_TCP_PORT=53\",\"KUBE_DNS_PORT_53_TCP_ADDR=10.96.0.10\",\"KUBERNETES_SERVICE_HOST=10.96.0.1\",\"KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443\",\"KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1\",\"KUBE_DNS_SERVICE_PORT=53\",\"KUBE_DNS_PORT_9153_TCP_PROTO=tcp\",\"KUBE_DNS_PORT_9153_TCP_ADDR=10.96.0.10\",\"KUBERNETES_PORT_443_TCP_PORT=443\",\"KUBE_DNS_SERVICE_PORT_DNS=53\",\"KUBE_DNS_PORT_9153_TCP=tcp://10.96.0.10:9153\",\"KUBE_DNS_PORT_53_UDP_PROTO=udp\",\"KUBE_DNS_PORT_53_TCP=tcp://10.96.0.10:53\",\"KUBE_DNS_PORT_53_TCP_PROTO=tcp\",\"KUBERNETES_PORT=tcp://10.96.0.1:443\",\"KUBERNETES_PORT_443_TCP_PROTO=tcp\",\"KUBE_DNS_SERVICE_HOST=10.96.0.10\",\"KUBE_DNS_PORT=udp://10.96.0.10:53\",\"KUBE_DNS_PORT_53_UDP_ADDR=10.96.0.10\",\"KUBE_DNS_PORT_9153_TCP_PORT=9153\",\"KUBERNETES_SER
...
```

### cri-o 

```
$ make ig
$ make minikube-start-cri-o
$ docker cp ig minikube-cri-o:/bin/ig
$ minikube ssh 
$ sudo ig list-containers -o json
...
    "ociConfig": "{\n\t\"ociVersion\": \"1.0.2-dev\",\n\t\"process\": {\n\t\t\"user\": {\n\t\t\t\"uid\": 0,\n\t\t\t\"gid\": 0,\n\t\t\t\"additionalGids\": [\n\t\t\t\t0\n\t\t\t]\n\t\t},\n\t\t\"args\": [\n\t\t\t\"/storage-provisioner\"\n\t\t],\n\t\t\"env\": [\n\t\t\t\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\",\n\t\t\t\"TERM=xterm\",\n\t\t\t\"HOSTNAME=minikube-cri-o\",\n\t\t\t\"KUBE_DNS_SERVICE_PORT_DNS=53\",\n\t\t\t\"KUBE_DNS_PORT=udp://10.96.0.10:53\",\n\t\t\t\"KUBE_DNS_PORT_53_TCP=tcp://10.96.0.10:53\",\n\t\t\t\"KUBE_DNS_PORT_53_TCP_PROTO=tcp\",\n\t\t\t\"KUBE_DNS_PORT_53_TCP_ADDR=10.96.0.10\",\n\t\t\t\"KUBE_DNS_PORT_9153_TCP_ADDR=10.96.0.10\",\n\t\t\t\"KUBERNETES_SERVICE_HOST=10.96.0.1\",\n\t\t\t\"KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443\",\n\t\t\t\"KUBERNETES_PORT_443_TCP_PORT=443\",\n\t\t\t\"KU[\n\t\t\t\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\",\n\t\t\t\"TERM=xterm\",\n\t\t\t\"HOSTNAME=minikube-cri-o\",\n\t\t\t\"KUBE_DNS_SERVICE_PORT_DNS=53\",\n\t\t\t\"KUBE_DNS_PORT=udp://10.96.0.10:53\",\n\t\t\t\"KUBE_DNS_PORT_53_TCP=tcp://10.96.0.10:53\",\n\t\t\t\"KUBE_DNS_PORT_53_TCP_PROTO=tcp\",\n\t\t\t\"KUBE_DNS_PORT_53_TCP_ADDR=10.96.0.10\",\n\t\t\t\"KUBE_DNS_PORT_9153_TCP_ADDR=10.96.0.10\",\n\t\t\t\"KUBERNETES_SERVICE_HOST=10.96.0.1\",\n\t\t\t\"KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443\",\n\t\t\t\"KUBERNETES_PORT_443_TCP_PORT=443\",\n\t\t\t\"KU
...
```

Also, integration tests has been adopted to verify OciConfig isn't empty for initial containers.

### CODE REVIEW

File: pkg/container-collection/options.go [hunk 1]
```diff
@@ -788,6 +799,63 @@ func WithOCIConfigEnrichment() ContainerCollectionOption {
...
+	return func(cc *ContainerCollection) error {
+		for _, container := range cc.initialContainers {
+			ppid := 0
+			sPath := filepath.Join(host.HostProcFs, fmt.Sprintf("%d", container.ContainerPid()), "status")
+			sFile, err := os.Open(sPath)
+			if err != nil {
+				log.Errorf("OCIConfig enricher: failed to open status file for container %s: %s", container.Runtime.ContainerID, err)
+				continue
+			}
+			defer sFile.Close()
```
Comment (alban) [inline pkg/container-collection/options.go:814]: If there are many containers in `cc.initialContainers`, the open file descriptors will accumulate and the defer will close all of them at the end.

I wonder if a util function is preferable to avoid this.

File: pkg/container-collection/options.go [hunk 2]
```diff
@@ -788,6 +799,63 @@ func WithOCIConfigEnrichment() ContainerCollectionOption {
...
+					log.Warnf("cannot parse PPid from /proc/PID/status: %s", err)
+				}
+				break
+			}
+			target, err := os.Readlink(filepath.Join(host.HostProcFs, fmt.Sprintf("%d", ppid), "cwd"))
+			if err != nil {
+				log.Errorf("OCIConfig enricher: failed to read cwd symlink for container %s: %s", container.Runtime.ContainerID, err)
+				continue
+			}
+			bPath := filepath.Join(host.HostRoot, target)
```
Comment (alban) [inline pkg/container-collection/options.go:840]:
```suggestion
			bPath := securejoin.SecureJoin(host.HostRoot, target)
```
But I think the `securejoin.SecureJoin(host.HostRoot, ...)` has to be done after appending `container.Runtime.ContainerID` and `config.json` with `filepath.Join`.

File: pkg/container-collection/options.go [hunk 3]
```diff
@@ -788,6 +799,63 @@ func WithOCIConfigEnrichment() ContainerCollectionOption {
...
+			defer sFile.Close()
+
+			// Read the status file to get the parent pid
+			reader := bufio.NewReader(sFile)
+			for {
+				line, err := reader.ReadString('\n')
+				if err != nil {
+					log.Errorf("OCIConfig enricher: failed to read status file for container %s: %s", container.Runtime.ContainerID, err)
+					break
+				}
```
Comment (alban) [inline pkg/container-collection/options.go:823]: Need to handle `io.EOF`.
reader.ReadString() can return `io.EOF` together with a non-empty `line`.

I wonder if ioutil.ReadFile is easier.

File: pkg/container-collection/options.go [hunk 4]
```diff
@@ -788,6 +792,61 @@ func WithOCIConfigEnrichment() ContainerCollectionOption {
...
+			}
+			bPath, err := filepath.EvalSymlinks(filepath.Join(host.HostProcFs, fmt.Sprintf("%d", ppid), "cwd"))
+			if err != nil {
+				log.Errorf("OCIConfig enricher: failed to read cwd for container %s: %s", container.Runtime.ContainerID, err)
+				continue
+			}
+
+			// In case of containerd, we get the bundle path of sandbox containers so we need to switch to the actual container directory.
+			if container.Runtime.RuntimeName == types.RuntimeNameContainerd && !strings.HasSuffix(bPath, container.Runtime.ContainerID) {
+				bPath = filepath.Join(filepath.Dir(bPath), container.Runtime.ContainerID)
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go:814]: > A good test would be to use containerd with an alternative state path configured in /etc/containerd/config.toml:

Good idea. I just tried with following and it works fine: 

```
$ docker@minikube-containerd:~$ grep state /etc/containerd/config.toml      
state = "/test/run/containerd"
$ docker@minikube-containerd:~$ sudo ls -la /test/run/containerd
total 20
drwx--x--x 5 root root 4096 May 27 16:43 .
drwx--x--x 3 root root 4096 May 27 16:43 ..
drwxr-xr-x 4 root root 4096 May 27 16:43 io.containerd.grpc.v1.cri
drwx--x--x 2 root root 4096 May 27 16:43 io.containerd.runtime.v1.linux
drwx--x--x 3 root root 4096 May 27 16:43 io.containerd.runtime.v2.task
```

Having said that, it was failing for `kubectl debug node ...` but works fine after `host.HostRoot`.

File: pkg/container-collection/options.go (hunk 3 above)
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go]: I adopted the implementation, Let me know if it looks good now!

File: pkg/container-collection/options.go [hunk 5]
```diff
@@ -781,6 +793,74 @@ func WithOCIConfigEnrichment() ContainerCollectionOption {
...
+	sPath := filepath.Join(host.HostProcFs, fmt.Sprintf("%d", container.ContainerPid()), "stat")
+	s, err := os.ReadFile(sPath)
+	if err != nil {
+		return 0, fmt.Errorf("reading stat file for container %s: %w", container.Runtime.ContainerID, err)
+	}
+	fields := strings.Fields(string(s))
+	if len(fields) < 4 {
+		return 0, fmt.Errorf("stat file for container %s has unexpected format: %s", container.Runtime.ContainerID, string(s))
+	}
+	ppid, err := strconv.Atoi(fields[3])
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go]: I based this PR on https://github.com/inspektor-gadget/inspektor-gadget/pull/4446 so we don't have the parsing issue!

File: pkg/container-collection/process.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+func (procOpts) WithVmSize() bool           { return false }
+func (procOpts) WithVmRSS() bool            { return false }
+func (procOpts) WithMemoryRelative() bool   { return false }
+func (procOpts) WithThreadCount() bool      { return false }
+func (procOpts) WithStartTime() bool        { return false }
+
+func (procOpts) TotalMemory() uint64              { return 0 }
+func (procOpts) NumCPU() int                      { return 0 }
+func (procOpts) LastCPUTime(_ int) (uint64, bool) { return 0, false }
+func (procOpts) BootTime() time.Time              { return time.Time{} }
```
Comment (alban) [inline pkg/container-collection/process.go:38]: I think the API in processhelpers.go can be improved so that consumers of the API don't have to define such a big interface.

But that could be done later

**Comment by mqasimsarfraz** [comment]
Thanks for the review!
